File: c++/698. Partition to K Equal Sum Subsets.cpp
class Solution {
public:
	bool canPartitionKSubsets(vector<int>& nums, int k) {
		int sum = accumulate(nums.begin(), nums.end(), 0);
		int n = nums.size();
		if (n < k || sum < k || sum % k != 0) return false;
		vector<int> vis(n, 0);

		return dfs(0, k, sum / k, sum / k, nums, vis);
	}

	bool dfs(int idx, int k, int len, int cur_len, vector<int> &nums, vector<int> &vis) {
		if (k == 0) return true;
		if (cur_len == 0)
			return dfs(0, k - 1, len, len, nums, vis);

		for (int i = idx; i < nums.size(); i++) {
			if (vis[i] == 1)
				continue;
			vis[i] = 1;
			if (nums[i] <= cur_len && dfs(i + 1, k, len, cur_len - nums[i], nums, vis))
				return true;
			vis[i] = 0;
		}
		return false;
	}
};

File: go/src/leetcode/698/698. Partition to K Equal Sum Subsets.go
/**
@author ZhengHao Lou
Date    2022/09/20
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/partition-to-k-equal-sum-subsets/
*/
func canPartitionKSubsets(nums []int, k int) bool {
	n := len(nums)
	var tot int
	for i := range nums {
		tot += nums[i]
	}
	if tot%k != 0 {
		return false
	}
	var g = tot / k
	var vis = make([]bool, n)
	var dfs func(idx, s, c int) bool
	dfs = func(idx, s, c int) bool {
		if c == 0 {
			return true
		}
		if s == 0 {
			return dfs(0, g, c-1)
		}

		for i := idx; i < n; i++ {
			if vis[i] {
				continue
			}

			vis[i] = true
			if s >= nums[i] && dfs(i+1, s-nums[i], c) {
				return true
			}
			vis[i] = false
		}
		return false
	}

	return dfs(0, g, k)
}

func main() {
	//fmt.Println(canPartitionKSubsets([]int{4, 3, 2, 3, 5, 2, 1}, 4))
	//fmt.Println(canPartitionKSubsets([]int{1, 2, 3, 4}, 3))
	fmt.Println(canPartitionKSubsets([]int{129, 17, 74, 57, 1421, 99, 92, 285, 1276, 218, 1588, 215, 369, 117, 153, 22}, 3))
}
